packages/flutter/lib/src/material/combo_box.dart

```diff
@@ -0,0 +1,747 @@
+  final InputDecorationTheme? inputDecorationTheme;
+
+  /// The [MenuStyle] that defines the visual attributes of the menu.
+  ///
+  /// The default width of the menu is set to the width of the text field.
+  final MenuStyle? menuStyle;
+
+  /// Descriptions of the menu items in the [ComboBox].
+  ///
+  /// This is a required parameter.
+  final List<ComboBoxEntry> comboBoxEntries;
+
+  @override
+  State<ComboBox> createState() => _ComboBoxState();
+}
+
+class _ComboBoxState extends State<ComboBox> {
+  final MenuController _controller = MenuController();
+  final GlobalKey _anchorKey = GlobalKey();
+  final GlobalKey _leadingKey = GlobalKey();
+  final FocusNode _textFocusNode = FocusNode();
+  final TextEditingController _textEditingController = TextEditingController();
+  late bool _enableFilter;
+  late List<ComboBoxEntry> filteredEntries;
+  List<Widget>? _initialMenu;
+  int? currentHighlight;
+  double? leadingPadding;
+  bool _menuHasEnabledItem = false;
+
+  @override
+  void initState() {
+    super.initState();
+    _enableFilter = widget.enableFilter;
+    filteredEntries = widget.comboBoxEntries;
+    for (final ComboBoxEntry entry in filteredEntries) {
+      if (entry.enabled) {
+        _menuHasEnabledItem = true;
+        break;
+      }
+    }
+    refreshLeadingPadding();
+  }
+
+  @override
+  void didUpdateWidget(ComboBox oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.comboBoxEntries != widget.comboBoxEntries) {
+      _menuHasEnabledItem = false;
+      for (final ComboBoxEntry entry in filteredEntries) {
+        if (entry.enabled) {
+          _menuHasEnabledItem = true;
+          break;
+        }
+      }
+    }
+    if (oldWidget.leadingIcon != widget.leadingIcon) {
+      refreshLeadingPadding();
+    }
+  }
+
+  void refreshLeadingPadding() {
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      setState(() {
+        leadingPadding = getWidth(_leadingKey);
+      });
+    });
+  }
+
+  double? getWidth(GlobalKey key) {
+    final BuildContext? context = key.currentContext;
+    if (context != null) {
+      final RenderBox box = context.findRenderObject()! as RenderBox;
+      return box.size.width;
+    }
+    return null;
+  }
+
+  List<ComboBoxEntry> filter(List<ComboBoxEntry> entries, TextEditingController textEditingController) {
+    final List<ComboBoxEntry> filteredEntries = entries.where((ComboBoxEntry entry)
+      => entry.label.toLowerCase().contains(textEditingController.text.toLowerCase())).toList();
+
+    return filteredEntries;
+  }
+
+  int? search(List<ComboBoxEntry> entries, TextEditingController textEditingController) {
+    for (int i = 0; i < entries.length; i++) {
+      final ComboBoxEntry entry = entries[i];
+      if (textEditingController.value.text.isNotEmpty
+          && entry.label.toLowerCase().contains(textEditingController.text.toLowerCase())) {
+        return i;
+      }
+    }
+
+    return null;
```

Comment:
A little simpler, and also doesn't check the controller's text every iteration:

```suggestion
    final String searchText = textEditingController.value.text.toLowerCase();
    if (searchText.isEmpty) {
      return null;
    }
    final int index = entries.indexWhere((ComboBox entry) => entry.label.toLowerCase().contains(searchText));
    return index != -1 ? index : null;
```